gluten-core/src/main/scala/io/glutenproject/execution/BroadcastNestedLoopJoinExecTransformer.scala

```diff
@@ -0,0 +1,206 @@
+        case RightOuter => right.outputPartitioning
+        case x =>
+          throw new IllegalArgumentException(
+            s"BroadcastNestedLoopJoin should not take $x as the JoinType with building left side")
+      }
+    case BuildRight =>
+      joinType match {
+        case _: InnerLike => left.outputPartitioning
+        case LeftOuter => left.outputPartitioning
+        case x =>
+          throw new IllegalArgumentException(
+            s"BroadcastNestedLoopJoin should not take $x as the JoinType with building right side")
+      }
+  }
+
+  override def doTransform(context: SubstraitContext): TransformContext = {
+    val streamedPlanContext = streamedPlan.asInstanceOf[TransformSupport].doTransform(context)
+    val (inputStreamedRelNode, inputStreamedOutput) =
+      (streamedPlanContext.root, streamedPlanContext.outputAttributes)
+
+    val buildPlanContext = buildPlan.asInstanceOf[TransformSupport].doTransform(context)
+    val (inputBuildRelNode, inputBuildOutput) =
+      (buildPlanContext.root, buildPlanContext.outputAttributes)
+
+    val expressionNode = condition.map {
+      expr =>
+        ExpressionConverter
+          .replaceWithExpressionTransformer(expr, inputStreamedOutput ++ inputBuildOutput)
+          .doTransform(context.registeredFunction)
+    }
+    val extensionNode =
+      JoinUtils.createExtensionNode(inputStreamedOutput ++ inputBuildOutput, validation = false)
+    val operatorId = context.nextOperatorId(this.nodeName)
+
+    val crossRel = RelBuilder.makeCrossRel(
+      inputStreamedRelNode,
+      inputBuildRelNode,
+      substraitJoinType,
+      expressionNode.orNull,
+      extensionNode,
+      context,
+      operatorId
+    )
+
+    val projectRelPostJoinRel = JoinUtils.createProjectRelPostJoinRel(
+      needSwitchChildren,
+      joinType,
+      inputStreamedOutput,
+      inputBuildOutput,
+      context,
+      operatorId,
+      crossRel,
+      inputStreamedOutput,
+      inputBuildOutput
+    )
+
+    JoinUtils.createTransformContext(
+      needSwitchChildren,
+      output,
+      projectRelPostJoinRel,
+      inputStreamedOutput,
+      inputBuildOutput)
+  }
+
+  override protected def doValidateInternal(): ValidationResult = {
+    if (!BackendsApiManager.getSettings.supportBroadcastNestedLoopJoinExec()) {
+      return ValidationResult.notOk("Broadcast Nested Loop join is not supported in this backend")
+    }
+    if (substraitJoinType == CrossRel.JoinType.UNRECOGNIZED) {
+      return ValidationResult.notOk(s"$joinType is not supported with Broadcast Nested Loop Join")
+    }
+    (joinType, buildSide) match {
+      case (LeftOuter, BuildLeft) | (RightOuter, BuildRight) =>
+        return ValidationResult.notOk(s"$joinType join is not supported with $buildSide")
+      case _ => // continue
+    }
+    val substraitContext = new SubstraitContext
+    val expressionNode = condition.map {
+      expr =>
+        ExpressionConverter
```

Comment:
This code is duplicate, can we move it in separate function and use in this and the above method